<img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/Practicum_AI_Logo.white_outline.svg?raw=1' width=250 alt='Practicum AI logo'> <img src='https://github.com/PracticumAI/practicumai.github.io/blob/main/images/icons/practicumai_python.png?raw=true' align='right' width=50>
***

# Data Analysis with pandas
In this module, you will learn about one of the most popular and valuable Python libraries for data analysis and manipulation: ***pandas***!

<img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/pandas_logo.png?raw=1' alt='The pandas logo' align='center' width=200>

### After this module, students will be able to:
* Explain the various components of a pandas DateFrame.
* Use `read_csv()` to create new DateFrames from data.
* Subset data with pandas index and DateFrame management functions.
* Use basic statistical functions to summarize a dataset.
* Use matplotlib functions to generate clear graphs and plots.


## <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/get_started_icon.svg?raw=1' alt="Let's get started header" width=40 align=center> Let's Get Started!

As we learned in the last module, we must **import** the pandas library before we can use it.
* We'll also import the scientific computing library NumPy for illustration purposes.
* NumPy and pandas are usually imported with abbreviated aliases, `np` and `pd`.

In [5]:
import numpy as np
import pandas as pd

## pandas Dataframes

pandas provides one of the most useful data structures for managing data: the DataFrame.

**DataFrames:**

* Allow fast, flexible, and efficient data manipulation.
* Are the natural data structure for **tabular** data.
* Are two-dimensional, containing both **rows** and **columns**.
* Consist of one or more **Series** (as columns).


***pandas*** is built on top of the **NumPy** library, which in practice means that most of the methods defined for NumPy Arrays apply to pandas Series/DataFrames.

**What makes pandas so useful for working with data?**
1. Powerful and efficient interface to access and manipulate individual records.
2. Tools for handling missing values and explicit values for missing data (`NaN` and `None`).
3. Relational database-style operations between DataFrames.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/note_icon.svg?raw=1' alt="Note icon" width=40 align=center> Note
> Some of the examples used in this notebook are different from those used in the corresponding video.

## Creating a Series

Let's start by creating our first pandas Series! (Remember: in the context of a **DataFrame** (table), each column is represented by a **Series**).

First, let's create some example data to illustrate the creation of our Series. in this example, we will use data that relates a person's "SpO2" reading from a pulse oximeter (which shows the percentage of oxygen in someone's blood), and their risk of adverse outcomes.

In [6]:
spo2 = [88, 99, 88, 85, 100, 98] # Review: this is a Python list, containing 6 values.
risk = ['high', 'low', 'high', 'high', 'low', 'low']

In [7]:
# Create our first Series.
spo2_series = pd.Series(spo2)
spo2_series # If a variable name is the last line in a code cell, Jupyter will display the contents of the variable.

0     88
1     99
2     88
3     85
4    100
5     98
dtype: int64

In [8]:
# Do the same for risk.
pd.Series(risk) # Display the 'risk' Series

0    high
1     low
2    high
3    high
4     low
5     low
dtype: object

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/alert_icon.svg?raw=1' alt="Alert icon" width=40 align=center> Alert!
> An important point to remember is that Python is "zero-indexed." That means that the first item in a list, or the first row in a table, has the index of 0.
>
> Notice the output above. The first column shows the Series index, and the second column has the value. The indices start at 0!
>
> This applies to lists, strings, arrays, and almost everything in Python. Keep this in mind--a common error is mistakenly using index 1, thinking you are getting the first item, where you need to use index 0.


## Creating a pandas Dataframe

* Now that we've seen a Series, we can see how these are combined as **columns** in a **DataFrame**.
* By popular convention, we will name our DataFrame variable `df`.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
> Create a small DataFrame from our example data.

In [9]:
df = pd.DataFrame({'spo2': spo2, 'risk': risk}) # Create the DataFrame
df # View the DataFrame

spo2  risk
0    88  high
1    99   low
2    88  high
3    85  high
4   100   low
5    98   low

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
> Create a larger DataFrame from random data.

In [10]:
# Make some random numbers for data.
# This makes an array of random integers from 1-10, then reshapes it into a
# 500x4 matrix (500 rows and 4 columns).
random_data = np.random.randint(1,10, size=2000).reshape(500, 4)

# Create some example column names
column_names = ['columnA', 'columnB', 'columnC', 'columnD']

# Create the DateFrame from the random data we generated.
# Since the input matrix has 500 rows and 4 columns, the DataFrame will have
# 500 rows and 4 columns.
df = pd.DataFrame(data=random_data, columns=column_names)
df # View the DataFrame.

columnA  columnB  columnC  columnD
0          5        4        9        2
1          9        2        7        3
2          1        5        2        3
3          8        7        9        5
4          9        8        5        3
..       ...      ...      ...      ...
495        3        6        5        1
496        4        7        2        1
497        5        7        1        8
498        6        3        1        1
499        1        1        5        2

[500 rows x 4 columns]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/note_icon.svg?raw=1' alt="Note icon" width=40 align=center> Note
> When viewing a large DataFrame in a Jupyter notebook, sometimes the data will be **truncated** as above.

We often want to see the first (or last) several rows of a DateFrame. We can use the `head()` and `tail()` functions.

In [11]:
# View the 1st 5 rows.
df.head()

columnA  columnB  columnC  columnD
0        5        4        9        2
1        9        2        7        3
2        1        5        2        3
3        8        7        9        5
4        9        8        5        3

In [12]:
# View the last 3 rows (head also takes a number of rows as an optional
# argument).
df.tail(3)

columnA  columnB  columnC  columnD
497        5        7        1        8
498        6        3        1        1
499        1        1        5        2

We can also limit our views to particular columns using the column names. If the column names have no spaces or special characters, we can use the "dot" notation: `df.columnA`. Otherwise, we must pass a list of columns: `df[['columnA', 'columnB']]`.

In [13]:
# View multiple columns or if there are spaces in names
# Note the double square brackets!
df[['columnA', 'columnB']].head()

columnA  columnB
0        5        4
1        9        2
2        1        5
3        8        7
4        9        8

# Creating a DataFrame from an existing data file

* Instead of creating a DataFrame from scratch (tedious!), we usually create a DataFrame by loading an existing data file.
* The most common file format for storing datasets is the **comma-separated value** `(.csv)` file.

> To create a pandas DataFrame from a `.csv` file, we'll use the `read_csv` function.

### Let's load some data

The main input argument to the `read_csv` function can be a local filename or a remote file URL. In this example, we'll be using an online file.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/note_icon.svg?raw=1' alt="Note icon" width=40 align=center> Note
> For information on how to upload a file into Jupyter, see the [handout](handouts/uploading-and-loading-a-CSV.pdf) (in the handouts folder). This works for single, relatively small files (under 1 or 2GB).

The dataset used is from the [GapMinder Project](https://www.gapminder.org/) and includes information on GDP, life expectancy, and population in various countries for different years.

In [14]:
url='https://raw.githubusercontent.com/PracticumAI/python-for-ai/main/data/gapminder_all.csv' # Direct link to the dataset on GitHub
df = pd.read_csv(url) # Create the DataFrame from the given URL
df # View the DateFrame we just created

continent         country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0      Africa         Algeria     2449.008185     3013.976023     2550.816880   
1      Africa          Angola     3520.610273     3827.940465     4269.276742   
2      Africa           Benin     1062.752200      959.601080      949.499064   
3      Africa        Botswana      851.241141      918.232535      983.653976   
4      Africa    Burkina Faso      543.255241      617.183465      722.512021   
..        ...             ...             ...             ...             ...   
137    Europe     Switzerland    14734.232750    17909.489730    20431.092700   
138    Europe          Turkey     1969.100980     2218.754257     2322.869908   
139    Europe  United Kingdom     9979.508487    11283.177950    12477.177070   
140   Oceania       Australia    10039.595640    10949.649590    12217.226860   
141   Oceania     New Zealand    10556.575660    12247.395320    13175.678000   

     gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0       3246.991771     4182.663766     4910.416756     5745.160213   
1       5522.776375     5473.288005     3008.647355     2756.953672   
2       1035.831411     1085.796879     1029.161251     1277.897616   
3       1214.709294     2263.611114     3214.857818     4551.142150   
4        794.826560      854.735976      743.387037      807.198586   
..              ...             ...             ...             ...   
137    22966.144320    27195.113040    26982.290520    28397.715120   
138     2826.356387     3450.696380     4269.122326     4241.356344   
139    14142.850890    15895.116410    17428.748460    18232.424520   
140    14526.124650    16788.629480    18334.197510    19477.009280   
141    14463.918930    16046.037280    16233.717700    17632.410400   

     gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0       5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1       2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2       1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3       6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4        912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   
..              ...  ...         ...         ...         ...         ...   
137    30281.704590  ...   5666000.0   6063000.0   6401400.0   6316424.0   
138     5089.043686  ...  29788695.0  33411317.0  37492953.0  42404033.0   
139    21664.787670  ...  53292000.0  54959000.0  56079000.0  56179000.0   
140    21888.889030  ...  10794968.0  11872264.0  13177000.0  14074100.0   
141    19007.191290  ...   2488550.0   2728150.0   2929100.0   3164900.0   

       pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0    20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1     7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2     3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3      970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4     6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  
..          ...         ...         ...         ...       ...       ...  
137   6468126.0   6649942.0   6995447.0   7193761.0   7361757   7554661  
138  47328791.0  52881328.0  58179144.0  63047647.0  67308928  71158647  
139  56339704.0  56981620.0  57866349.0  58808266.0  59912431  60776238  
140  15184200.0  16257249.0  17481977.0  18565243.0  19546792  20434176  
141   3210650.0   3317166.0   3437674.0   3676187.0   3908037   4115771  

[142 rows x 38 columns]

We can see that this dataset contains 142 rows and 38 columns corresponding to each country's GDP, life expectancy, and population between 1952 and 2007.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/note_icon.svg?raw=1' alt="Note icon" width=40 align=center> Note
> Stop for a second here. What happened to the variable `df` we defined for our SpO2 data?


## Viewing parts of a Dataframe
* Dataframes often hold many rows (in many AI projects, you might be dealing with hundreds of thousands, or even millions, of samples!)
* Sometimes, we want a preview or a sense of the data stored in our DataFrame.
* In the below examples, we'll see a few ways to view parts of a DataFrame.
    * This includes viewing a subset of columns, a subset of rows, or both.

With so many columns, it can be a bit hard to see all of them, but we can get a list of the columns in the DataFrame with the `.columns` method:

In [15]:
df.columns

Index(['continent', 'country', 'gdpPercap_1952', 'gdpPercap_1957',
       'gdpPercap_1962', 'gdpPercap_1967', 'gdpPercap_1972', 'gdpPercap_1977',
       'gdpPercap_1982', 'gdpPercap_1987', 'gdpPercap_1992', 'gdpPercap_1997',
       'gdpPercap_2002', 'gdpPercap_2007', 'lifeExp_1952', 'lifeExp_1957',
       'lifeExp_1962', 'lifeExp_1967', 'lifeExp_1972', 'lifeExp_1977',
       'lifeExp_1982', 'lifeExp_1987', 'lifeExp_1992', 'lifeExp_1997',
       'lifeExp_2002', 'lifeExp_2007', 'pop_1952', 'pop_1957', 'pop_1962',
       'pop_1967', 'pop_1972', 'pop_1977', 'pop_1982', 'pop_1987', 'pop_1992',
       'pop_1997', 'pop_2002', 'pop_2007'],
      dtype='object')

Next, let's revisit viewing some of a DataFrame's **rows**. Since pandas is so flexible, there are several ways to accomplish this. The examples below use the pandas' functions `head` and `tail`.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 1

> Use `head` to view the first 5 rows of our `df` DataFrame. (**Hint:** Given a DataFrame `X`, calling `X.head()` will show the first `5` rows of the DataFrame.  

In [16]:
# View only the first 5 rows of the DateFrame.
# Your code here
df.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0     5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1     2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2     1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3     6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4      912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   

     pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0  20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1   7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2   3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3    970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4   6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  

[5 rows x 38 columns]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 2
> Use `head` to view the first **10** rows of `df`. (**Hint:** You can also provide an input argument `n` to the `head` function to show the first `n` rows (instead of the default `5` rows)).

In [17]:
# View only the first 10 rows of the DataFrame.
# Your code here
df.head(10)

continent                   country  gdpPercap_1952  gdpPercap_1957  \
0    Africa                   Algeria     2449.008185     3013.976023   
1    Africa                    Angola     3520.610273     3827.940465   
2    Africa                     Benin     1062.752200      959.601080   
3    Africa                  Botswana      851.241141      918.232535   
4    Africa              Burkina Faso      543.255241      617.183465   
5    Africa                   Burundi      339.296459      379.564628   
6    Africa                  Cameroon     1172.667655     1313.048099   
7    Africa  Central African Republic     1071.310713     1190.844328   
8    Africa                      Chad     1178.665927     1308.495577   
9    Africa                   Comoros     1102.990936     1211.148548   

   gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
0     2550.816880     3246.991771     4182.663766     4910.416756   
1     4269.276742     5522.776375     5473.288005     3008.647355   
2      949.499064     1035.831411     1085.796879     1029.161251   
3      983.653976     1214.709294     2263.611114     3214.857818   
4      722.512021      794.826560      854.735976      743.387037   
5      355.203227      412.977514      464.099504      556.103265   
6     1399.607441     1508.453148     1684.146528     1783.432873   
7     1193.068753     1136.056615     1070.013275     1109.374338   
8     1389.817618     1196.810565     1104.103987     1133.984950   
9     1406.648278     1876.029643     1937.577675     1172.603047   

   gdpPercap_1982  gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972  \
0     5745.160213     5681.358539  ...  11000948.0  12760499.0  14760787.0   
1     2756.953672     2430.208311  ...   4826015.0   5247469.0   5894858.0   
2     1277.897616     1225.856010  ...   2151895.0   2427334.0   2761407.0   
3     4551.142150     6205.883850  ...    512764.0    553541.0    619351.0   
4      807.198586      912.063142  ...   4919632.0   5127935.0   5433886.0   
5      559.603231      621.818819  ...   2961915.0   3330989.0   3529983.0   
6     2367.983282     2602.664206  ...   5793633.0   6335506.0   7021028.0   
7      956.752991      844.876350  ...   1523478.0   1733638.0   1927260.0   
8      797.908101      952.386129  ...   3150417.0   3495967.0   3899068.0   
9     1267.100083     1315.980812  ...    191689.0    217378.0    250027.0   

     pop_1977    pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  \
0  17152804.0  20033753.0  23254956.0  26298373.0  29072015.0  31287142   
1   6162675.0   7016384.0   7874230.0   8735988.0   9875024.0  10866106   
2   3168267.0   3641603.0   4243788.0   4981671.0   6066080.0   7026113   
3    781472.0    970347.0   1151184.0   1342614.0   1536536.0   1630347   
4   5889574.0   6634596.0   7586551.0   8878303.0  10352843.0  12251209   
5   3834415.0   4580410.0   5126023.0   5809236.0   6121610.0   7021078   
6   7959865.0   9250831.0  10780667.0  12467171.0  14195809.0  15929988   
7   2167533.0   2476971.0   2840009.0   3265124.0   3696513.0   4048013   
8   4388260.0   4875118.0   5498955.0   6429417.0   7562011.0   8835739   
9    304739.0    348643.0    395114.0    454429.0    527982.0    614382   

   pop_2007  
0  33333216  
1  12420476  
2   8078314  
3   1639131  
4  14326203  
5   8390505  
6  17696293  
7   4369038  
8  10238807  
9    710960  

[10 rows x 38 columns]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 3
> Use `tail` to view the **last** 5 rows of our `df` DataFrame. (**Hint:** Given a DateFrame `X`, calling `X.tail()` will show the **last** `5` rows of the DateFrame).  

In [18]:
# View the last 5 rows of the DataFrame.
# Your code here
df.tail()

continent         country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
137    Europe     Switzerland    14734.232750    17909.489730    20431.092700   
138    Europe          Turkey     1969.100980     2218.754257     2322.869908   
139    Europe  United Kingdom     9979.508487    11283.177950    12477.177070   
140   Oceania       Australia    10039.595640    10949.649590    12217.226860   
141   Oceania     New Zealand    10556.575660    12247.395320    13175.678000   

     gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
137    22966.144320     27195.11304    26982.290520    28397.715120   
138     2826.356387      3450.69638     4269.122326     4241.356344   
139    14142.850890     15895.11641    17428.748460    18232.424520   
140    14526.124650     16788.62948    18334.197510    19477.009280   
141    14463.918930     16046.03728    16233.717700    17632.410400   

     gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
137    30281.704590  ...   5666000.0   6063000.0   6401400.0   6316424.0   
138     5089.043686  ...  29788695.0  33411317.0  37492953.0  42404033.0   
139    21664.787670  ...  53292000.0  54959000.0  56079000.0  56179000.0   
140    21888.889030  ...  10794968.0  11872264.0  13177000.0  14074100.0   
141    19007.191290  ...   2488550.0   2728150.0   2929100.0   3164900.0   

       pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
137   6468126.0   6649942.0   6995447.0   7193761.0   7361757   7554661  
138  47328791.0  52881328.0  58179144.0  63047647.0  67308928  71158647  
139  56339704.0  56981620.0  57866349.0  58808266.0  59912431  60776238  
140  15184200.0  16257249.0  17481977.0  18565243.0  19546792  20434176  
141   3210650.0   3317166.0   3437674.0   3676187.0   3908037   4115771  

[5 rows x 38 columns]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 4
> Use `tail` to view the **last** `10` rows of `df`. (**Hint:** You can also provide an input argument `n` to the `tail` function to show the last `n` rows (instead of the default `5` rows)).

In [19]:
# View only the last 10 rows of the DateFrame.
# Your code here
df.tail(10)

continent          country  gdpPercap_1952  gdpPercap_1957  \
132    Europe           Serbia     3581.459448     4981.090891   
133    Europe  Slovak Republic     5074.659104     6093.262980   
134    Europe         Slovenia     4215.041741     5862.276629   
135    Europe            Spain     3834.034742     4564.802410   
136    Europe           Sweden     8527.844662     9911.878226   
137    Europe      Switzerland    14734.232750    17909.489730   
138    Europe           Turkey     1969.100980     2218.754257   
139    Europe   United Kingdom     9979.508487    11283.177950   
140   Oceania        Australia    10039.595640    10949.649590   
141   Oceania      New Zealand    10556.575660    12247.395320   

     gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
132     6289.629157     7991.707066    10522.067490    12980.669560   
133     7481.107598     8412.902397     9674.167626    10922.664040   
134     7402.303395     9405.489397    12383.486200    15277.030170   
135     5693.843879     7993.512294    10638.751310    13236.921170   
136    12329.441920    15258.296970    17832.024640    18855.725210   
137    20431.092700    22966.144320    27195.113040    26982.290520   
138     2322.869908     2826.356387     3450.696380     4269.122326   
139    12477.177070    14142.850890    15895.116410    17428.748460   
140    12217.226860    14526.124650    16788.629480    18334.197510   
141    13175.678000    14463.918930    16046.037280    16233.717700   

     gdpPercap_1982  gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972  \
132    15181.092700    15870.878510  ...   7616060.0   7971222.0   8313288.0   
133    11348.545850    12037.267580  ...   4237384.0   4442238.0   4593433.0   
134    17866.721750    18678.534920  ...   1582962.0   1646912.0   1694510.0   
135    13926.169970    15764.983130  ...  31158061.0  32850275.0  34513161.0   
136    20667.381250    23586.929270  ...   7561588.0   7867931.0   8122293.0   
137    28397.715120    30281.704590  ...   5666000.0   6063000.0   6401400.0   
138     4241.356344     5089.043686  ...  29788695.0  33411317.0  37492953.0   
139    18232.424520    21664.787670  ...  53292000.0  54959000.0  56079000.0   
140    19477.009280    21888.889030  ...  10794968.0  11872264.0  13177000.0   
141    17632.410400    19007.191290  ...   2488550.0   2728150.0   2929100.0   

       pop_1977    pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  \
132   8686367.0   9032824.0   9230783.0   9826397.0  10336594.0  10111559   
133   4827803.0   5048043.0   5199318.0   5302888.0   5383010.0   5410052   
134   1746919.0   1861252.0   1945870.0   1999210.0   2011612.0   2011497   
135  36439000.0  37983310.0  38880702.0  39549438.0  39855442.0  40152517   
136   8251648.0   8325260.0   8421403.0   8718867.0   8897619.0   8954175   
137   6316424.0   6468126.0   6649942.0   6995447.0   7193761.0   7361757   
138  42404033.0  47328791.0  52881328.0  58179144.0  63047647.0  67308928   
139  56179000.0  56339704.0  56981620.0  57866349.0  58808266.0  59912431   
140  14074100.0  15184200.0  16257249.0  17481977.0  18565243.0  19546792   
141   3164900.0   3210650.0   3317166.0   3437674.0   3676187.0   3908037   

     pop_2007  
132  10150265  
133   5447502  
134   2009245  
135  40448191  
136   9031088  
137   7554661  
138  71158647  
139  60776238  
140  20434176  
141   4115771  

[10 rows x 38 columns]

Now, let's see some ways to view a subset of a DateFrame's **columns**.

**Reminder:** If the last line inside a code cell is simply a variable name, Jupyter will display the value of that variable.

For a given DateFrame `df` and desired column `col`, one way to view a single column is to write `df['col']`. (**Important:** the column name `col` must be in quotes!).

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 5
> Use the above method to view the `gdpPercap_2007` column of our `df` DateFrame.

In [20]:
# View the gdpPercap_2007 column of the DateFrame.
# Your code here
df['gdpPercap_2007']

0       6223.367465
1       4797.231267
2       1441.284873
3      12569.851770
4       1217.032994
           ...     
137    37506.419070
138     8458.276384
139    33203.261280
140    34435.367440
141    25185.009110
Name: gdpPercap_2007, Length: 142, dtype: float64

We can also adapt this method to view multiple columns. Instead of passing in a single string column name `'C'`, we will pass in a **list** of string column names, e.g. `['A', 'B', 'C']`. For example: `df[['A', 'B', 'C']]`

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/note_icon.svg?raw=1' alt="Note icon" width=40 align=center> Note
> **Important:** Notice the double brackets! DataFrames use a similar syntax as lists. To get a single column, we can use the column's name: e.g., `df['A']` would give us column A. Here, we want multiple columns, so a list of column names is provided.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 6
> Use the above method to view the `gdpPercap_2007`, `lifeExp_2007`, and `pop_2007` columns of our `df` DateFrame.

In [22]:
# Display the columns `gdpPercap_2007`, `lifeExp_2007`, and `pop_2007`
# Your code here
df[['gdpPercap_2007','lifeExp_2007','pop_2007']]

gdpPercap_2007  lifeExp_2007  pop_2007
0       6223.367465        72.301  33333216
1       4797.231267        42.731  12420476
2       1441.284873        56.728   8078314
3      12569.851770        50.728   1639131
4       1217.032994        52.295  14326203
..              ...           ...       ...
137    37506.419070        81.701   7554661
138     8458.276384        71.777  71158647
139    33203.261280        79.425  60776238
140    34435.367440        81.235  20434176
141    25185.009110        80.204   4115771

[142 rows x 3 columns]

## Creating new columns

* We can easily create new columns in an existing DateFrame.
* When we use mathematical operations as below, they will be applied to every element of the Series.
* This way, we can create new columns from different combinations of existing ones.

Let's use an equation to calculate the change in GDP from 1962 to 2007.

In [23]:
df['gdp_change'] = (df['gdpPercap_1962'] - df['gdpPercap_2007'])
df.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1967    pop_1972    pop_1977    pop_1982  \
0     5681.358539  ...  12760499.0  14760787.0  17152804.0  20033753.0   
1     2430.208311  ...   5247469.0   5894858.0   6162675.0   7016384.0   
2     1225.856010  ...   2427334.0   2761407.0   3168267.0   3641603.0   
3     6205.883850  ...    553541.0    619351.0    781472.0    970347.0   
4      912.063142  ...   5127935.0   5433886.0   5889574.0   6634596.0   

     pop_1987    pop_1992    pop_1997  pop_2002  pop_2007    gdp_change  
0  23254956.0  26298373.0  29072015.0  31287142  33333216  -3672.550585  
1   7874230.0   8735988.0   9875024.0  10866106  12420476   -527.954525  
2   4243788.0   4981671.0   6066080.0   7026113   8078314   -491.785809  
3   1151184.0   1342614.0   1536536.0   1630347   1639131 -11586.197794  
4   7586551.0   8878303.0  10352843.0  12251209  14326203   -494.520973  

[5 rows x 39 columns]

## Deleting columns

We can also **drop** (delete) columns.

Let's remove the `gdp_change` column that we previously created.

In [24]:
df.drop(columns=['gdp_change']) # Will this remove the column?
df.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1967    pop_1972    pop_1977    pop_1982  \
0     5681.358539  ...  12760499.0  14760787.0  17152804.0  20033753.0   
1     2430.208311  ...   5247469.0   5894858.0   6162675.0   7016384.0   
2     1225.856010  ...   2427334.0   2761407.0   3168267.0   3641603.0   
3     6205.883850  ...    553541.0    619351.0    781472.0    970347.0   
4      912.063142  ...   5127935.0   5433886.0   5889574.0   6634596.0   

     pop_1987    pop_1992    pop_1997  pop_2002  pop_2007    gdp_change  
0  23254956.0  26298373.0  29072015.0  31287142  33333216  -3672.550585  
1   7874230.0   8735988.0   9875024.0  10866106  12420476   -527.954525  
2   4243788.0   4981671.0   6066080.0   7026113   8078314   -491.785809  
3   1151184.0   1342614.0   1536536.0   1630347   1639131 -11586.197794  
4   7586551.0   8878303.0  10352843.0  12251209  14326203   -494.520973  

[5 rows x 39 columns]

**What happened? `gdp_change` is still there!**

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/tip_icon.svg?raw=1' alt="Tip icon" width=40 align=center> Tip
> Unless you specify `inplace=True` or assign the DateFrame back to itself (or a different variable name), **the `drop` method does not change the original DateFrame!**

Below is one way to permanently remove a column from a DateFrame by assigning the result to a new DateFrame variable `df2`.

In [25]:
# After dropping the column, we assign the resulting DateFrame to a new variable.
# This will persist the change.
df2 = df.drop(columns=['gdp_change'])
df2.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0     5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1     2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2     1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3     6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4      912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   

     pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0  20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1   7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2   3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3    970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4   6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  

[5 rows x 38 columns]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 7

> Create a new DateFrame called `df3`, that contains everything in the `df` DateFrame **except** the `continent` column.

In [26]:
# Code it!
df3= df.drop(columns=['continent'])
df3.head()

country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0       Algeria     2449.008185     3013.976023     2550.816880   
1        Angola     3520.610273     3827.940465     4269.276742   
2         Benin     1062.752200      959.601080      949.499064   
3      Botswana      851.241141      918.232535      983.653976   
4  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  gdpPercap_1992  ...    pop_1967    pop_1972    pop_1977  \
0     5681.358539     5023.216647  ...  12760499.0  14760787.0  17152804.0   
1     2430.208311     2627.845685  ...   5247469.0   5894858.0   6162675.0   
2     1225.856010     1191.207681  ...   2427334.0   2761407.0   3168267.0   
3     6205.883850     7954.111645  ...    553541.0    619351.0    781472.0   
4      912.063142      931.752773  ...   5127935.0   5433886.0   5889574.0   

     pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  \
0  20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216   
1   7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476   
2   3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314   
3    970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131   
4   6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203   

     gdp_change  
0  -3672.550585  
1   -527.954525  
2   -491.785809  
3 -11586.197794  
4   -494.520973  

[5 rows x 38 columns]

## Descriptive Statistics

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/alert_icon.svg?raw=1' alt="Alert icon" width=40 align=center> Alert!
> Before diving into an AI project, becoming familiar with the data you will be modeling is critical!  pandas has an extensive collection of optimized methods designed to help you analyze your data.

A preliminary data discovery and exploration phase can help you better understand the data you will be modeling and has many downstream advantages:
* You may identify potential problems with the dataset (e.g., missing values, outliers, incomplete columns, irrelevant information).
* Results of data exploration may help you decide which preprocessing or modeling approach to take.
* A better understanding of your dataset will help you interpret future results.

#### Remember Python functions?
pandas offers many **functions** that take as **input** a set of values (example: all of the values from the `gdpPercap_2007` column) and **return** a calculated **output** value based on which function was used.

Here are a few examples (most are self-explanatory!):
* `sum` (the sum of the values)
* `count` (the number of values), `value_counts` (the number unique values)
* `min` (the minimum value), `max` (the maximum value)
* `mean` (the mean value), `median` (the median value), `mode` (the mode of the values)
* `var` (the variance of the values), `std` (the standard deviation of the values)
* `quantile` (the quantiles of the values)

Let's take a look at a few examples using our dataset!

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
Here's how we can compute the **mean** GDP in 2007:

In [27]:
# Compute the mean GDP in 2007
df['gdpPercap_2007'].mean()

np.float64(11680.071819878167)

It's that easy! Let's compute a few more descriptive statistics:

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 8
> Compute the **minimum** GDP in 2007.

In [28]:
# Compute the minimum GDP in 2007
# Your code here

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 9
> Compute the **maximum** GDP in 2007.

In [ ]:
# Compute the maximum GDP in 2007
# Your code here

Finally, the `describe` function can be used on an entire DateFrame to display several standard statistics computed for each column. (**Note:** Be aware of each column's values and data types, as sometimes the statistics will not make sense.)

In [ ]:
# (Almost) everything, (almost) everywhere, all at once
# Note: will only compute statistics for columns with numerical data.
df.describe()

## Filtering data
pandas makes it easy to select, analyze, and process subsets of your data based on some criteria, `C`. This is accomplished with the syntax: `df[C]`, which will return a subset of DateFrame `df` that **matches** the criteria `C` listed inside brackets `[]`.

### Comparing values
In pandas, we can create filtering criteria using familiar mathematical operators to compare values:
* `>`: greater than
* `<`: less than
* `>=`: greater than or equal to
* `<=`: less than or equal to
* `!=`: is not equal to
* `==`: is equal to a single value
* **Bonus:** `.isin(L)`: equals any value in the list `L`

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
Display 2007 GDP is **greater** than $40,000.
* In this example, the criterion `C` is `df['gdpPercap_2007'] > 40000`.
* Recall the pandas filtering syntax: `df[C]`.
* So, this is what we want: `df[df['gdpPercap_2007'] > 40000]`.

In [ ]:
# Select rows where 2007 GDP is greater than $40,000
df[df['gdpPercap_2007'] > 40000]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 10
> Display where 2007 GDP is **less** than $500

In [ ]:
# Display rows where 2007 GDP is less than $500

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
Display the first 5 rows for African countries.

* Recall that the symbol `==` will check for equality (we want the continent to be equal to the single value `Africa`).
* In this example, the criterion `C` is `df['continent] == 'Africa'`.
* Given the pandas filtering syntax: `df[C]`, this is what we want: `df[df['continent'] == 'Africa']`.

In [ ]:
# Display the first 5 rows for African countries.
df[df['continent'] == 'Africa'].head()

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 11
> Display the first 5 rows for European countries (`'continent'`==`'Europe'`).

In [ ]:
# Display the first 5 rows for European countries ('continent'=='Europe').
# Write your code here


### Combining multiple criteria

Multiple criteria can be combined using either the `&` or `|` symbols:
* C1 `&` C2: This will select rows that match **both** criterion C1 **and** criterion C2.
* C1 `|` C2: This will select rows that match **either** criterion C1 **or** criterion C2.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
Display rows where 2007 GDP is greater than or equal to 30,000 and 2007 life expectancy is under 60. In this example, we have two criteria:
* C1: `df['gdpPercap_2007'] >= 30000`
* C2: `df['lifeExp_2007'] < 60`

In [ ]:
# Select rows where 2007 GDP is greater than or equal to 30,000 and 2007 life expectancy is under 60.
df[(df['gdpPercap_2007'] >= 30000) & (df['lifeExp_2007'] < 60)]

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 12
> Display all European countries that have 2007 life expectancies over 80.

In [ ]:
# Display all European countries that have 2007 life expectancies over 80.
# Your code here



### Analyzing filtered data
We can also perform our descriptive statistics on a filtered subset!

In [ ]:
df[(df['continent'] == 'Europe') & (df['lifeExp_2007'] > 80)].describe()

## Data visualization

Let's conclude this lesson by learning how pandas can enhance our understanding of a dataset with its many built-in visualization tools.

Let's see the data! This time, we'll load our dataset into a DateFrame variable named `X`.

> ![A hand icon](https://github.com/paigewill-sys/python-for-ai/blob/main/images/hand.png?raw=1) Wait! `X`?? Didn't we say to avoid single-letter variable names and generally use lowercase?? Well...rules are...there to be broken when there is a good reason. In much of machine learning and AI, we work with datasets where we want to make a model to generate a prediction. Part of the goal is to take input data and predict an output or outcome. Thinking of a graph with `x`- and `y`-values, we can think of the `x`-value as an input and the `y`-value as the outcome. So, `x` and `y` make sense as variable names when thinking about values on a graph and datasets more generally. Going further, we borrow a convention from math, where a matrix (or a table with several columns) is referenced with a capital letter. Since our DataFrame has multiple columns, we will use the capital X for the name.
>
> We also want to ensure you understand that while we frequently use `df` for a DataFrame variable name, `df` is just a name, and nothing is special about it.

In [ ]:
X = pd.read_csv(url) # url is defined above
X.head()

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
Let's visually examine the distribution of 2007 GDPs with the pandas `hist` (histogram) function.

In [ ]:
hist = X['gdpPercap_2007'].hist()

We can modify many parameters of the `hist` function to create a more informative figure.

In [ ]:
hist = X['gdpPercap_2007'].hist(bins=72, grid=False)
hist.set_title('Distribution of 2007 GDP of Countries')
hist.set_ylabel('Count')
hist.set_xlabel('GDP in 2007')
display(hist)

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 13
> Visualize the distribution of the 2007 life expectancies.

In [ ]:
# Visualize the distribution of the 2007 life expectancies.
# Your code here


This can also be written as: `X.hist(['lifeExp_2007'], bins=40)`

That syntax makes it easier to plot multiple histograms side-by-side. Below, we compare the distributions of life expectancies in 1962 and 2007.

In [ ]:
X.hist(column=['lifeExp_1962','lifeExp_2007'], bins=40, sharey=True, sharex=True)

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/example_icon.svg?raw=1' alt="Example icon" width=40 align=center> Example
We can use the `df.plot.box` function to create box plots from columns in our DateFrame:

In [ ]:
X.plot.box(column='lifeExp_2007', by='continent')

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 14
> Practice with the `hist` and `box` functions by exploring different columns in our DateFrame. Experiment with data filtering methods to refine your figures further. **Hint:** We can see all available columns with `print(X.columns)`.

In [ ]:
### Experiment with code!


## Data visualization with `plotnine`

[`plotnine` is a popular data visualization library](https://plotnine.readthedocs.io/) that implements the Grammar of Graphics paradigm and [ggplot](https://ggplot2.tidyverse.org/) syntax in Python. For more information and exercises using `plotnine`, check out the [04_more_data_visualization.ipynb](04_more_data_visualization.ipynb) notebook.

Here, we want to demonstrate how the plots created above using `pandas` can also be made using `plotnine`.

In [ ]:
import plotnine as p9

In [ ]:
# plotnine implementation of the first histogram from above.
(p9.ggplot(X, p9.aes(x='gdpPercap_2007'))
 + p9.geom_histogram()
)

Not a bad start.

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 15

When you ran the code above, you should have gotten a warning:`PlotnineWarning: 'stat_bin()' using 'bins = 8'. Pick better value with 'binwidth'.`

Use the [`plotnine` documentation](https://plotnine.readthedocs.io/) to:

* Change the `binwidth` to another value you think would be good.
* Change the color of the bars to something more colorful.
* Change the x-axis to "Gross Domestic Product Per Capita, 2007."
* Make any other changes you want to make to this histogram.


And here's the boxplot we made above.

In [ ]:
(p9.ggplot(X)
    + p9.geom_boxplot(p9.aes(x='continent', y='lifeExp_2007'))
 )

### <img src='https://github.com/paigewill-sys/python-for-ai/blob/main/images/exercise_icon.svg?raw=1' alt="Exercise icon" width=40 align=center> Exercise 16

Again, this is an excellent start. Explore `plotnine` to see what you can do to improve this!

***
#### Attribution
Some content in this learning experience was adapted from Ben Shickel's [AI for Medicine bootcamp](https://github.com/gatorai/scripps) content.  